libfwupdplugin/fu-firmware-attributes.c

```diff
@@ -0,0 +1,156 @@
+#include "fu-path.h"
+#include "fu-string.h"
+
+static gchar *
+fu_firmware_attributes_build(const gchar *name, GError **error)
+{
+	g_autofree gchar *sysfsfwdir = fu_path_from_kind(FU_PATH_KIND_SYSFSDIR_FW_ATTRIB);
+	g_autofree gchar *tmp = NULL;
+	const gchar *current_value;
+	const gchar *driver;
+	GDir *dir;
+
+	g_return_val_if_fail(name != NULL, NULL);
+
+	dir = g_dir_open(sysfsfwdir, 0, error);
+	if (dir == NULL)
+		return NULL;
+
+	driver = g_dir_read_name(dir);
+	if (driver == NULL) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_FOUND,
+			    "failed to find driver in %s",
+			    sysfsfwdir);
+		return NULL;
+	}
+
+	if (g_dir_read_name(dir)) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_FOUND,
+			    "Unable to support multiple drivers found in %s",
+			    sysfsfwdir);
+		return NULL;
+	}
+
+	current_value = g_strcmp0(name, "pending_reboot") != 0 ? "current_value" : NULL;
+	tmp = g_build_filename(sysfsfwdir, driver, "attributes", name, current_value, NULL);
+	if (!g_file_test(tmp, G_FILE_TEST_EXISTS)) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_FOUND,
+			    "failed to find attribute %s",
+			    tmp);
+		return NULL;
+	}
+
+	return g_steal_pointer(&tmp);
+}
+
+/**
+ * fu_firmware_attribute_has_attr:
+ * @name: A firmware attribute name found in /sys/class/firmware-attributes/$DRIVER/attributes
+ *
+ * Determines if the system offers a given firmware attribute name
+ *
+ * Since: 1.8.4
+ **/
+gboolean
+fu_firmware_attribute_has_attr(const gchar *name)
+{
+	g_autofree gchar *tmp = fu_firmware_attributes_build(name, NULL);
+	return (tmp != NULL);
+}
+
+/**
+ * fu_firmware_attributes_get_value:
+ * @name: A firmware attribute name found in /sys/class/firmware-attributes/$DRIVER/attributes
+ * @error: (nullable): optional return location for an error
+ *
+ * Gets the string stored in an attribute
+ *
+ * Since: 1.8.4
+ **/
+gchar *
+fu_firmware_attributes_get_value(const gchar *name, GError **error)
+{
+	g_autofree gchar *path = fu_firmware_attributes_build(name, error);
+	g_autofree gchar *buf = NULL;
+
+	if (path == NULL)
+		return NULL;
+
+	if (!g_file_get_contents(path, &buf, NULL, error)) {
+		g_prefix_error(error, "failed to read attribute from %s: ", path);
+		return NULL;
+	}
+
+	return g_steal_pointer(&buf);
+}
+
+/**
+ * fu_firmware_attributes_get_pending:
+ * @result: (out): Whether a reboot is pending
+ * @error: (nullable): optional return location for an error
+ *
+ * Determines if the system will apply changes to attributes upon reboot
+ *
+ * Since: 1.8.4
+ **/
+gboolean
+fu_firmware_attributes_get_pending(gboolean *result, GError **error)
+{
+	g_autofree gchar *buf = NULL;
+	guint64 val = 0;
+
+	g_return_val_if_fail(result != NULL, FALSE);
+
+	buf = fu_firmware_attributes_get_value("pending_reboot", error);
+	if (!buf)
+		return FALSE;
+	if (!fu_strtoull(buf, &val, 0, G_MAXUINT32, error))
+		return FALSE;
+
+	*result = val > 0;
```

Comment:
is this more tightly defined by the kernel? i.e. should this be val & 0x01 for instance

Reply:
fixed